### DISCUSSION TITLE
Make `vet` to warn about unneeded Option/Result return type of a function

### DISCUSSION THREAD

**Discussion by ArtemkaKun** [vlang/v#18620] [Ideas]
### Describe the feature

I think it's a great idea to pinpoint unneeded Option/Result usage with a warning or even to make `fmt` to remove them

### Use Case

```v
pub fn clean_punctuation(string_to_clean string) !string {
	mut punctuation_regex := common.create_regex(punctuation.punctuation_regex_filter)

	return punctuation_regex.replace(string_to_clean, '')
}
```

This method doesn't need to Result return type.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### Version used

V 0.3.2 161847e

### Environment details (OS name and version, etc.)

```
OS: linux, "Garuda Linux"
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 3800X 8-Core Processor
CC version: cc (GCC) 12.2.0

getwd: /home/yuart/Projects/text-processing
vmodules: /home/yuart/.vmodules
vroot: /usr/lib/vlang
vexe: /usr/lib/vlang/v
vexe mtime: 2022-12-01 13:45:41
is vroot writable: false
is vmodules writable: true
V full version: V 0.3.2 161847e
```

**Comment by spytheman** [comment]
The return type of a function is part of its signature, and that in turn is like a promise about what the function expects, and what it will produce. The programmer is solely responsible for writing that, and the compiler should not try to second guess it, just check whether all callers conform to that promise/contract, and whether the body of the function conforms to it too (returning a string, when the return type is int is not allowed etc).

Besides, vfmt has only local information about the current .v file, and can not know how a function is used by all its callers, so it can not make that change without breaking code.

Such a feature could be a part of a linter tool like `v vet`, but not part of the compiler or vfmt.

**Comment by ArtemkaKun** [comment]
Agreed. I will transfer request to a `v vet` then.

I think it can just display an error like "Your function X declares !string return type, but all callers use string type only. Consider to change X return type to string"

**Comment by JalonSolov** [comment]
I think it should be the other way around.  "Your function declares !string type, but only returns string."  That would be a logic error in the function.  The callers should already be warned if they don't have `or` blocks or propagate with `!` after the call.
